x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java

```diff
@@ -481,59 +482,33 @@ private static Limit descendantLimit(UnaryPlan unary) {
         }
     }
 
-    static class DuplicateLimitAfterMvExpand extends OptimizerRules.OptimizerRule<Limit> {
+    private static boolean isTransparentForLimit(LogicalPlan plan) {
+        return plan instanceof Eval || plan instanceof Project || plan instanceof RegexExtract || plan instanceof Enrich;
+    }
 
+    protected static class PushDownMvExpandPastOrderBy extends OptimizerRules.OptimizerRule<OrderBy> {
```

Comment:
I tried to reconstruct the behavior of this rule and the reasoning behind it, and I have to say that it took me some time and I'm still unsure I got all the nuances.

This is a problem by itself: the rule tries to analyze a potentially large portion of the plan and then makes changes in positions that are very far from the entry point of the rule; it's hard to understand now, so it will be even harder to maintain in the long term, when the original intentions will be harder to recollect and when the language (and the planner) will be richer and more complex.

This said, I tried to understand if the rule is correct in all cases, and _my conclusion is that it's not_.
My understanding is that the intention is the following: if you have a plan with `sort | mv_expand | ... | sort`, you replace it with `mv_expand | sort | ... | sort`, in the assumption that eventually the first `sort` will be removed (because redundant?).
This is correct as long as there are no commands in between that drop possible candidate results, or for which the order actually matters.

The case that we are ignoring now is LIMIT: a query like `sort | mv_expand |  ... | limit | ... | sort` is different from `mv_expand | sort |  ... | limit | ... | sort`, especially when the first sort happens on the expanded key. Both `descendantMvExpand()` and `replaceChildUntilMvExpandAndOrderBy()` are not taking this case into consideration.

Cases we could miss tomorrow are all the commands that rely on the intermediate order, that include inline stats, commands that access previous records and other cases we can hardly completely define now.

In short, this rule would be much better defined as a local rule that converts `sort | mv_expand` -> `mv_expand | sort`, but unfortunately it's not generally valid, and defining all the cases when it's valid is risky and error prone. Also, the general usefulness of this change is not completely clear (to me at least).
We can try to fix the above, but we are paying a high cost to maintainability for arguably small advantages.

Reply:
++, agree that it's hard to verify this rule for all cases, esp. as we might add new operators that might invalidate it in the future, silently.

If we want to keep the `sort | mv_expand | ... | sort -> mv_expand | sort | ... | sort`, we probably should restrict the operators that can go in the ` ... ` part explicitly, ideally adding at least a unit test per such operator to verify that output doesn't change after optimization. That is a lot of work, though :/